src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -2174,43 +2207,18 @@ private AssemblySymbol GetForwardedToAssembly(string fullName, int arity, Diagno
                 while (current != null);
             }
 
-            // NOTE: This won't work if the type isn't using CLS-style generic naming (i.e. `arity), but this code is
-            // only intended to improve diagnostic messages, so false negatives in corner cases aren't a big deal.
-            var metadataName = MetadataTypeName.FromFullName(fullName, useCLSCompliantNameArityEncoding: true, forcedArity: arity);
-
-            var containingAssembly = this.Compilation.Assembly;
+            var metadataName = MetadataHelpers.ComposeAritySuffixedMetadataName(name, arity);
 
-            // This method is only called after lookup has failed, so the containing (source!) assembly can't
-            // have a forwarder to another assembly.
-            NamedTypeSymbol forwardedType = null;
-            foreach (var referencedAssembly in containingAssembly.Modules[0].GetReferencedAssemblySymbols())
+            var fullName = qualifierOpt != null && !ReferenceEquals(qualifierOpt, Compilation.GlobalNamespace) ? qualifierOpt.ToDisplayString(SymbolDisplayFormat.QualifiedNameOnlyFormat) + "." + metadataName : metadataName;
```

Comment:
>ToDisplayString [](start = 126, length = 15)

ToDisplayString is still being used to construct a metadata name. Please use a helper from `MetadataTypeName` instead.

Reply:
I didn't see your comments before, sorry.

ToDisplayString is not the right helper for making a metadata name - that is for human display and may be more expensive than necessary. If `MetadataTypeName FromNamespaceAndTypeName` does what you need, please use it.

It is possible to have a type imported from metadata that uses a metadata name that does not use the convention of adding a backtick and the arity to the name. Such a type won't work with this code. So it is not safe to assume that "we" are using arity naming without documenting that assumption. I'm just suggesting simply adding back the comment explaining that such a type won't work properly here (thus documenting the assumption).

Reply:
Ok, that makes sense to me now. Will change it accordingly!

Reply:
So, I looked into this, but I don't really understand how my use of `ToDisplayString` can be replaced by any of the provided `MetadataTypeName` helper methods. I'm using `ToDisplayString` to get a string for the namespace I have as a `NamespaceOrTypeSymbol`. But all the `MetadataTypeName` helper methods take strings as input, so I still need to convert the `NamespaceOrTypeSymbol` to a string somehow to use any of them. I could also use `ToString`, but that just calls `ToDisplayString`. And `NamespaceOrTypeSymbol .Name` will not give me the parent namespaces?

Reply:
@gafter could you help me with what is the correct API to use here?